### PR TITLE
Add Petitions Dashboard widget

### PR DESCRIPTION
**Pull request by bahiirwa**
Issue https://github.com/ClassicPress/ClassicPress/issues/40 discussions lead to a feature request of a dashboard widget that shares the petitions suggested by the community.

This PR seeks to add that to core.

## How has this been tested?
Initially this was developed as a plugin https://github.com/bahiirwa/ClassicPressPetitionsDashboard and tested with both alpha-1 and beta-1versions before testing in core commits on this branch.

## Screenshots

screenshot 2018-12-04 14 04 53

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/wp-admin/includes/dashboard.php [hunk 1]
```diff
@@ -1113,7 +1117,8 @@ function wp_dashboard_events_news() {
 		<?php
 			printf(
 				'<a href="%1$s" target="_blank">%2$s <span class="screen-reader-text">%3$s</span><span aria-hidden="true" class="dashicons dashicons-external"></span></a>',
-				'https://www.classicpress.net/blog/',
+				/* translators: If a Rosetta site exists (e.g. https://es.wordpress.org/news/), then use that. Otherwise, leave untranslated. */
```
Comment (dikiyforester) [inline src/wp-admin/includes/dashboard.php]: This overrides previous PR #248. You'll probably need to fetch the latest changes first.

File: src/wp-admin/includes/dashboard.php [hunk 2]
```diff
@@ -1517,41 +1567,278 @@ function wp_dashboard_empty() {}
...
+
+	/**
+	 * Response should be an array with:
+	 *  'most-wanted []' - array() - A user-friendly platform name, if it can be determined
+	 *  'trending []' - array() - A user-friendly browser name
+	 *  'recent []' - array() - The version of the browser the user is using
+	 *  'tags [] - array() - The version of the browser the user is using
+	 *  'link - string - The version of the browser the user is using
+	 */
+	$response = wp_remote_get( $api_url );
```
Comment (dikiyforester) [inline src/wp-admin/includes/dashboard.php]: Do I correctly understand that this happens on page load?
This probably should be similar to `wp_dashboard_primary()`, which load news using AJAX request and doesn't affect the page loading time.

File: src/wp-admin/includes/dashboard.php (hunk 1 above)
Comment (bahiirwa) [reply src/wp-admin/includes/dashboard.php]: Rebase done.

File: src/wp-admin/includes/dashboard.php (hunk 2 above)
Comment (bahiirwa) [reply src/wp-admin/includes/dashboard.php]: Need to look into this implementation. However, it's on page load

File: src/wp-admin/css/dashboard.css [hunk 3]
```diff
@@ -1,1316 +1,1418 @@
 #wpbody-content #dashboard-widgets.columns-1 .postbox-container {
-	width: 100%;
```
Comment (dikiyforester) [inline src/wp-admin/css/dashboard.css]: You probably need to use tabs instead of spaces for line indentions.

File: src/wp-admin/js/dashboard.js [hunk 4]
```diff
@@ -2,498 +2,579 @@
 var ajaxWidgets, ajaxPopulateWidgets, quickPressLoad;
 window.wp = window.wp || {};
 
-jQuery(document).ready( function($) {
```
Comment (dikiyforester) [inline src/wp-admin/js/dashboard.js]: Same issue with line indentions.

File: src/wp-admin/includes/dashboard.php [hunk 5]
```diff
@@ -5,7 +5,6 @@
  * @package ClassicPress
  * @subpackage Administration
  */
-
```
Comment (dikiyforester) [inline src/wp-admin/includes/dashboard.php]: I'm not sure about these deleted empty lines. Seems like unnecessary changes.

File: src/wp-admin/includes/dashboard.php [hunk 6]
```diff
@@ -1555,3 +1401,244 @@ function wp_welcome_panel() {
 </div>
 	<?php
 }
+
+/**
+ * Returns HTML Mark up for the Petitions Dashboard Widget
+ *
+ * @since CP-1.0.0
```
Comment (nylen) [inline src/wp-admin/includes/dashboard.php]: Here and elsewhere, please remove the `CP-` prefix and just write `@since 1.0.0`.  We've prefixed WP versions with `WP-` so that it's clear they are from before ClassicPress started, but we don't need to do this for ClassicPress versions.

File: src/wp-admin/includes/dashboard.php (hunk 2 above)
Comment (nylen) [reply src/wp-admin/includes/dashboard.php]: This API endpoint is designed to be callable directly from within a browser.

An alternative would be to cache the result of the query in a transient.  However, calling the API from the browser would be preferable because it avoids issues with failed requests and how they interact with caching.

File: src/wp-admin/css/dashboard.css (hunk 3 above)
Comment (nylen) [reply src/wp-admin/css/dashboard.css]: Yes, there are multiple files overwritten / reformatted here.  Possibly a broken editor configuration.

Comment (bahiirwa) [reply src/wp-admin/css/dashboard.css]: I have a linter which I need to turn off

Comment (bahiirwa) [reply src/wp-admin/css/dashboard.css]: I realized from GitHub, that it uses 8 spaces for tab. I am adjusting that in my editor.

File: src/wp-admin/includes/dashboard.php (hunk 2 above)
Comment (bahiirwa) [reply src/wp-admin/includes/dashboard.php]: I have added a transient key for the JSON data from URL if no prior key is found. It has an expiry after 12 hours.

Comment (bahiirwa) [reply src/wp-admin/includes/dashboard.php]: That reduces the API calls and also the load times on page load.

File: src/wp-admin/includes/dashboard.php (hunk 5 above)
Comment (bahiirwa) [reply src/wp-admin/includes/dashboard.php]: Fixed

File: src/wp-admin/js/dashboard.js (hunk 4 above)
Comment (bahiirwa) [reply src/wp-admin/js/dashboard.js]: Fixed

**Comment by bahiirwa** [comment]
Travis finally green. Transients set up to handle the API calls to 12 hours intervals.

File: src/wp-admin/includes/dashboard.php [hunk 7]
```diff
@@ -1555,3 +1558,251 @@ function wp_welcome_panel() {
...
+
+		/**
+		 * Response should be an array with:
+		 *  'most-wanted []' - array() - A user-friendly platform name, if it can be determined
+		 *  'trending []' - array() - A user-friendly browser name
+		 *  'recent []' - array() - The version of the browser the user is using
+		 *  'tags [] - array() - The version of the browser the user is using
+		 *  'link - string - The version of the browser the user is using
+		 */
+		$response = wp_remote_get( $api_url );
```
[the remark this answers is not in the document]
Comment (nylen) [reply src/wp-admin/includes/dashboard.php]: @dikiyforester I think this current approach is OK for a first pass.  Using AJAX also has some drawbacks, so it is not a clear trade-off.  With AJAX, the initial page load time is slightly faster, but the time to load the content is always slightly longer.  This way the content would also depend on JavaScript to appear at all.

File: src/wp-admin/includes/dashboard.php [hunk 8]
```diff
@@ -1555,3 +1558,251 @@ function wp_welcome_panel() {
...
+
+		if ( ! is_array( $response ) ) {
+			return;
+		}
```
Comment (nylen) [inline src/wp-admin/includes/dashboard.php]: How about a more explicit error check here, something like this:

```
if ( is_wp_error( $response ) || 200 !==  wp_remote_retrieve_response_code( $response ) ) {
    return;
}
```

File: src/wp-admin/includes/dashboard.php [hunk 9]
```diff
@@ -1555,3 +1558,251 @@ function wp_welcome_panel() {
...
+
+		if ( is_array( $response ) ) {
+			$dashboard_petitions_results = $response['body']; // use the content
+		}
```
Comment (nylen) [inline src/wp-admin/includes/dashboard.php]: This `is_array` check is redundant.

File: src/wp-admin/includes/dashboard.php [hunk 10]
```diff
@@ -1555,3 +1558,251 @@ function wp_welcome_panel() {
...
+		set_transient( 'dashboard_petitions_results', $dashboard_petitions_results, 12 * HOUR_IN_SECONDS );
+	}
+	
+	/** 
+	 * Parse URL Ouput into PHP array
+	 * 
+	 * @since 1.0.0
+	 * @return array
+	 */
+	$json = json_decode( $dashboard_petitions_results, true );
```
Comment (nylen) [inline src/wp-admin/includes/dashboard.php]: This should probably happen _before_ setting the transient, and the result should be checked (an invalid response will cause a JSON parse error, which we should not store in the transient).

There is also a typo `Ouput` in the comment, but I don't think this comment is necessary at all.

File: src/wp-admin/includes/dashboard.php [hunk 11]
```diff
@@ -1555,3 +1558,251 @@ function wp_welcome_panel() {
...
+	 * 'votes' - Count Up Voted of petition
+	 * 'link' - URL to the petitions website page 
+	 * 'title' - Title of Petition 
+	 * 'author' - Name of Petition Lead 
+	 * 'status' - Status of petition  
+	 * 'createdTime' - Timestamp of the petition request 
+	 */
+	$most_wanted = $json['most-wanted']; // get most wanted (upvoted)
+	$recent = $json['recent']; // get recent
+	$trending = $json['trending']; // get trending
```
Comment (nylen) [inline src/wp-admin/includes/dashboard.php]: We can remove these variables...

File: src/wp-admin/includes/dashboard.php [hunk 12]
```diff
@@ -1555,3 +1558,251 @@ function wp_welcome_panel() {
...
+						$link = $value['link'];
+						$title = $value['title'];
+						$author = $value['createdBy'];
+						$status = $value['status'];
+						$createdTime = $value['createdAt'];
+
+						cp_dashboard_petitions_table_body ( $votesCount, $link, $title, $author, $status, $createdTime );
+
+					}
+				}
```
Comment (nylen) [inline src/wp-admin/includes/dashboard.php]: ... and instead of repeating this section of code 3 times, we can use `$json[ $list_item ]`.

Comment (nylen) [reply src/wp-admin/includes/dashboard.php]: I would also rename this data array from `$json` to `$petitions`.  This should fit well with the changes to move `json_decode` to before the `set_transient` call.

File: src/wp-admin/js/dashboard.js [hunk 13]
```diff
@@ -496,4 +496,35 @@ jQuery( function( $ ) {
...
+		
+		var clickedTab = event.currentTarget;
+		var anchor = event.target;
+		var activePaneID = anchor.getAttribute('href');
+		
+		// Get petitions-pane with same ID & add .active
+		clickedTab.classList.add('active');
+		$('#'+ activePaneID).addClass('active');
+	}
+});
```
Comment (nylen) [inline src/wp-admin/js/dashboard.js]: As far as the basic structure of this functionality, there are two ways to go:

1. Load the petitions data from the browser using an AJAX request to the API endpoint.  Make sure an appropriate message is shown with JavaScript disabled.
2. Load the petitions data on the server using `wp_remote_get`.  Make sure the tabs work with JavaScript disabled.

A benefit of using approach (2) is that the functionality *can* work using progressive enhancement with JavaScript enabled.  So, let's take advantage of this benefit.  It can work something like this:

- The active tab is determined by `$_GET['petitions_tab']`.  The CSS classes for active tab and hidden/visible content are assigned appropriately (and `empty( $_GET['petitions_tab'] )` means the first tab is shown).
- The tabs are structured using `<a>` elements which point to URLs like `?petitions_tab=TABNAME` so that they work without JavaScript enabled.
- The JavaScript overrides these default URLs to show a new content pane and mark a different tab as active.

**Comment by nylen** [comment]
Some smaller review points here, which I think can be discussed and addressed in the individual comment threads above.

The biggest thing that's left to do is to make sure the new functionality does something reasonable with JavaScript disabled.  Given the way it is structured, it's possible to keep it fully working under this situation.

From https://developer.wordpress.org/themes/advanced-topics/javascript-best-practices/#standard-javascript:

> Ensure your site still works without JavaScript first – then add JavaScript to provide additional capabilities. This is a form of Progressive Enhancement.

Regardless of whether WordPress itself has adhered to this guideline in recent versions, it is still a good guideline.

File: src/wp-admin/includes/dashboard.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (nylen) [reply src/wp-admin/includes/dashboard.php]: > Anyway, I would add my vision in a parallel branch, just let me know if it worth.

Maybe a PR against this branch to change over to an AJAX-based implementation?  I'm not sure how much of this code you would be able to reuse though.

Semi-related:  https://github.com/ClassicPress/ClassicPress/pull/259#discussion_r239663891

**Comment by dikiyforester** [comment]
> From https://developer.wordpress.org/themes/advanced-topics/javascript-best-practices/#standard-javascript:
> 
> > Ensure your site still works without JavaScript first – then add JavaScript to provide additional capabilities. This is a form of Progressive Enhancement.
> 
> Regardless of whether WordPress itself has adhered to this guideline in recent versions, it is still a good guideline.

I think the main idea behind this guideline, is that site itself shouldn't rely on JavaScript. In our case, we have a Petitions widget which is secondary and doesn't affect site functionality anyhow (and shouldn't affect site load as well).

File: src/wp-admin/includes/dashboard.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (dikiyforester) [reply src/wp-admin/includes/dashboard.php]: > Maybe a PR against this branch to change over to an AJAX-based implementation? I'm not sure how much of this code you would be able to reuse though.

Sure, we'll get something to compare.

File: src/wp-admin/includes/dashboard.php (hunk 8 above)
Comment (bahiirwa) [reply src/wp-admin/includes/dashboard.php]: Should we add a UX friendly line? Some thing about broken data?

File: src/wp-admin/includes/dashboard.php (hunk 9 above)
Comment (nylen) [reply src/wp-admin/includes/dashboard.php]: You're also checking `$response` here, not the response body.

File: src/wp-admin/includes/dashboard.php (hunk 8 above)
Comment (nylen) [reply src/wp-admin/includes/dashboard.php]: Sure, good idea.  This will come up when the server doesn't have a working internet connection, for example.

We should probably also set a timeout of around 5 seconds (at most) for this request.

**Comment by nylen** [comment]
Copy/pasting from [above](https://github.com/ClassicPress/ClassicPress/pull/259#discussion_r239663891) so it doesn't get lost when the PR is updated:

> As far as the basic structure of this functionality, there are two ways to go:
>
>1. Load the petitions data from the browser using an AJAX request to the API endpoint.  Make sure an appropriate message is shown with JavaScript disabled.
>2. Load the petitions data on the server using `wp_remote_get`.  Make sure the tabs work with JavaScript disabled.

I am fine with either of these two options **as a starting point** and I think we should merge whichever one is ready first, then follow up with a separate PR if desired.

File: src/wp-admin/includes/dashboard.php [hunk 14]
```diff
@@ -1555,3 +1558,251 @@ function wp_welcome_panel() {
...
+	 * 'title' - Title of Petition 
+	 * 'author' - Name of Petition Lead 
+	 * 'status' - Status of petition  
+	 * 'createdTime' - Timestamp of the petition request 
+	 */
+	$most_wanted = $json['most-wanted']; // get most wanted (upvoted)
+	$recent = $json['recent']; // get recent
+	$trending = $json['trending']; // get trending
+
+	/**
```
Comment (dikiyforester) [inline src/wp-admin/includes/dashboard.php]: @bahiirwa Would be good to configure your editor to automatically remove trailing spaces.

**Comment by dikiyforester** [comment]
@nylen please check out alternative PR #268 made using AJAX approach.

**Comment by nylen** [comment]
I've just merged #268 (which also includes many of the same commits from this PR).  Thanks @bahiirwa and @dikiyforester for your hard work on this feature!
